app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/PeerGroupHandshake.kt

```diff
@@ -0,0 +1,124 @@
+ * handshake between the two devices. The purpose of the handshake is to allow the file sending
+ * device to obtain the IP address of the file receiving device (When the file sending device
+ * is the wifi direct group owner, it doesn't have the IP address of the peer device by default).
+ *
+ * After obtaining the IP address, the sender also shares metadata regarding the file transfer
+ * (no. of files & their names) with the receiver.
+ */
+abstract class PeerGroupHandshake(private var groupInfo: WifiP2pInfo) {
+  private val HANDSHAKE_MESSAGE = "Request Kiwix File Sharing"
+  suspend fun handshake(): InetAddress? = withContext(Dispatchers.IO) {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Handshake in progress")
+    }
+    return@withContext when {
+      groupInfo.groupFormed && groupInfo.isGroupOwner && isActive -> {
+        groupOwnerReadHandShakeAndExchangeMetaData()
+      }
+      groupInfo.groupFormed && isActive -> {  // && !groupInfo.isGroupOwner
+        readHandShakeAndExchangeMetaData()
+      }
+      else -> return@withContext null
+    }
+  }
+
+  private fun readHandShakeAndExchangeMetaData(): InetAddress? {
+    try {
+      Socket().use { client ->
+        client.reuseAddress = true
+        client.connect(
+          InetSocketAddress(
+            groupInfo.groupOwnerAddress.hostAddress,
+            PEER_HANDSHAKE_PORT
+          ), 15000
+        )
+        val objectOutputStream = ObjectOutputStream(client.getOutputStream())
+        // Send message for the peer device to verify
+        objectOutputStream.writeObject(HANDSHAKE_MESSAGE)
+        exchangeFileTransferMetadata(client.getInputStream(), client.getOutputStream())
+        return@use groupInfo.groupOwnerAddress
+      }
+    } catch (ex: Exception) {
+      ex.printStackTrace()
+    }
+    return null
+  }
+
+  private fun groupOwnerReadHandShakeAndExchangeMetaData(): InetAddress? {
+    try {
+      ServerSocket(PEER_HANDSHAKE_PORT)
+        .use { serverSocket ->
+          serverSocket.reuseAddress = true
+          val server = serverSocket.accept()
+          val objectInputStream = ObjectInputStream(server.getInputStream())
+          val kiwixHandShakeMessage = objectInputStream.readObject()
+
+          // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+          return@groupOwnerReadHandShakeAndExchangeMetaData if (isKiwixHandshake(
+              kiwixHandShakeMessage
+            )
+          ) {
+            if (BuildConfig.DEBUG) {
+              Log.d(TAG, "Client IP address: " + server.inetAddress)
+            }
+            exchangeFileTransferMetadata(server.getInputStream(), server.getOutputStream())
+            server.inetAddress
+          } else {
+            // Selected device is not accepting wifi direct connections through the kiwix app
+            null
+          }
+        }
+    } catch (ex: Exception) {
+      ex.printStackTrace()
+    }
+    return null
```

Comment:
this only ever returns null, as with the method above return the try and move the null into the catch

Reply:
oh no it works, my eyes glazed over `return@groupOwnerReadHandShakeAndExchangeMetaData`, still do the above tho